runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,551 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+   // to pass the address of that value as an argument. We prepare the register below.
+   if (isWrite)
+      {
+      TR::Node *valueNode = node->getFirstChild();
+      if (isInstanceField)
+         {
+         // Pass in valueNode so it can be set to the correct node.
+         TR::TreeEvaluator::getIndirectWrtbarValueNode(cg, node, valueNode, false);
+         }
+      // First load the actual value into the register.
+      TR::Register *valueReferenceReg = cg->evaluate(valueNode);
+
+      TR::DataType nodeType = valueNode->getDataType();
+      TR::SymbolReference *sr = cg->allocateLocalTemp(nodeType);
+      TR::MemoryReference *valueMR = generateS390MemoryReference(valueNode, sr, cg);
+      if (valueReferenceReg->getKind() == TR_GPR)
+         {
+         // Use STG if the dataType is an uncompressed TR::Address or TR::Int64. ST otherwise.
+         auto mnemonic = TR::DataType::getSize(nodeType) == 8 ? TR::InstOpCode::STG : TR::InstOpCode::ST;
+         // Now store the value onto the stack.
+         generateRXInstruction(cg, mnemonic, node, valueReferenceReg, valueMR);
+         }
+      else if (valueReferenceReg->getKind() == TR_FPR)
+         {
+         auto mnemonic = nodeType == TR::Float ? TR::InstOpCode::STE : TR::InstOpCode::STD;
+         // Now store the value onto the stack.
+         generateRXInstruction(cg, mnemonic, node, valueReferenceReg, valueMR);
+         }
+      else
+         {
+         TR_ASSERT_FATAL(false, "Unsupported register kind (%d) for fieldwatch.", valueReferenceReg->getKind());
+         }
+      valueReferenceReg = cg->allocateRegister();
+
+      // Now load the memory location back into the register so that it can be used
+      // as an argument register for the VM helper call.
+      TR::MemoryReference *tempMR = generateS390MemoryReference(*valueMR, 0, cg);
+      generateRXInstruction(cg, TR::InstOpCode::LA, node, valueReferenceReg, tempMR);
+
+      dependencies->addPreCondition(valueReferenceReg, helperLink->getIntegerArgumentRegister(numArgs));
+      dependencies->addPostCondition(valueReferenceReg, helperLink->getIntegerArgumentRegister(numArgs));
+      valueReferenceReg->setPlaceholderReg();
+
+      cg->stopUsingRegister(valueReferenceReg);
+      }
+
+   // These registers will hold Entry Point and Return Address registers, which are required when generating a directCall.
+   TR::Register *scratch1 = cg->allocateRegister();
+   TR::Register *scratch2 = cg->allocateRegister();
+   dependencies->addPostCondition(scratch1, cg->getEntryPointRegister());
+   dependencies->addPostCondition(scratch2, cg->getReturnAddressRegister());
+
+   // Now generate the call to VM Helper to report the fieldwatch.
+   TR::Instruction *call = generateDirectCall(cg, node, false /*myself*/, cg->symRefTab()->findOrCreateRuntimeHelper(helperIndex, false, false, false), dependencies);
+   call->setNeedsGCMap(0x0000FFFF);
+   call->setDependencyConditions(dependencies);
+   cg->resetIsLeafMethod();
```

Comment:
I've never seen this API before. What is it's purpose?

Reply:
I thought we must do this when generating direct calls. But it looks like this is only used in the x86 codegen here:
https://github.com/eclipse/omr/blob/4c5291abe5606ef9a238328f3e2829baf1fb706a/compiler/x/codegen/X86SystemLinkage.cpp#L471-L479

And the setter is only ever called during initialization of the CG object:
https://github.com/eclipse/omr/blob/4c5291abe5606ef9a238328f3e2829baf1fb706a/compiler/codegen/OMRCodeGenerator.cpp#L270

I don't think we need to do this in the z codegen. I see no occurrences of this or the flag in the z codegen, and see no observable change in behaviour in my testing. So I've removed it here: 088b97c